library/ssl_tls13_client.c

```diff
@@ -3816,6 +3816,270 @@ static int ssl_hrr_postprocess( mbedtls_ssl_context* ssl,
+         return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "ticket->length: %u", ticket_len ) );
+
+    /* Ticket Extension */
+    ext_len = ( (size_t) buf[ i + ticket_len ] << 8 ) |
+              ( (size_t) buf[ i + ticket_len + 1 ] );
+
+    used += ext_len;
+
+    if( used != buflen )
+    {
+         MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad new session ticket message" ) );
+         return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
+    }
+
+    /* Check if we previously received a ticket already. */
+    if( ssl->session->ticket != NULL || ssl->session->ticket_len > 0 )
+    {
+        mbedtls_free( ssl->session->ticket );
+        ssl->session->ticket = NULL;
+        ssl->session->ticket_len = 0;
+    }
+
+    if( ( ticket = mbedtls_calloc( 1, ticket_len ) ) == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "ticket alloc failed" ) );
+        return( MBEDTLS_ERR_SSL_ALLOC_FAILED );
+    }
+
+    memcpy( ticket, buf + i, ticket_len );
+    i += ticket_len;
+    ssl->session->ticket = ticket;
+    ssl->session->ticket_len = ticket_len;
+
+    MBEDTLS_SSL_DEBUG_MSG( 4, ( "ticket->extension length: %u", ext_len ) );
+
+    i += 2;
+    /* We are not storing any extensions at the moment */
+    MBEDTLS_SSL_DEBUG_BUF( 3, "ticket->extension", &buf[i], ext_len );
+
+    /* Compute PSK based on received nonce and resumption_master_secret
+     * in the following style:
+     *
+     *  HKDF-Expand-Label( resumption_master_secret,
+     *                    "resumption", ticket_nonce, Hash.length )
+     */
+
+    suite_info = mbedtls_ssl_ciphersuite_from_id( ssl->session->ciphersuite );
+
+    if( suite_info == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+    }
+
+    hash_length = mbedtls_hash_size_for_ciphersuite( suite_info );
+
+    if( hash_length == -1 )
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+
+    MBEDTLS_SSL_DEBUG_BUF( 3, "resumption_master_secret",
+                           ssl->session->app_secrets.resumption_master_secret,
+                           hash_length );
+
+    /* Computer resumption key
+     *
+     *  HKDF-Expand-Label( resumption_master_secret,
+     *                    "resumption", ticket_nonce, Hash.length )
+     */
+    ret = mbedtls_ssl_tls1_3_hkdf_expand_label( suite_info->mac,
+                    ssl->session->app_secrets.resumption_master_secret,
+                    hash_length,
+                    MBEDTLS_SSL_TLS1_3_LBL_WITH_LEN( resumption ),
+                    ssl->session->ticket_nonce,
+                    ssl->session->ticket_nonce_len,
+                    ssl->session->key,
+                    hash_length );
+
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 2, "Creating the ticket-resumed PSK failed", ret );
+        return( ret );
+    }
+
+    ssl->session->resumption_key_len = hash_length;
```

Comment:
The naming should make it clear that `resumption_key_length` is the length of `key`. Could you align this?